### DISCUSSION TITLE
Multiple Characteristics with this UUID, refer to your desired characteristic by the `handle` attribute instead.

### DISCUSSION THREAD

**Discussion by ASabovic** [hbldh/bleak#939] [Resolved discussions]
Hi guys,

I enabled the BLE communication between my Arduino device and PC, and it works perfectly while my Arduino board is constantly connected to the BLE. Now, due to my project requirements, I need to turn off BLE on my board once I do not need it to use. When the data is ready to be sent again, I turn on BLE again and send the data to my PC. At the beginning of my experiments, everything works properly, but once when I disconnect my Arduino board and connect it again, I got this error message: 

"Multiple Characteristics with this UUID, refer to your desired characteristic by the `handle` attribute instead." It seems that my PC cannot recognize the Arduino board again and automatically it is not able to receive the message. Can you maybe help me to solve this issue? Btw, I always use the same service and characteristic UUID, so it's weird that the PC is confused even if it re-establishes the connection with the same device again. Thanks in advance!

Here is my code: 

```python
from cgi import test
import asyncio
from datetime import datetime
from aioconsole import ainput
import os
from typing import Any, Callable

from bleak import BleakClient

message_uuid = "2d2F88c4-f244-5a80-21f1-ee0224e80658"
address = "D2:AD:0B:29:4A:26"
data_buffer_size = 5120

class ArduinoConnection:

    client: BleakClient = None

    def __init__(
        self,
        loop: asyncio.AbstractEventLoop,
        message_uuid: str,
        data_buffer_cleaner: Callable[[str, Any], None],
        data_buffer_size: int
        #address: str
    ):
        self.loop = loop
        self.message_uuid = message_uuid
        self.data_buffer_cleaner = data_buffer_cleaner
        self.data_buffer_size = data_buffer_size
		
		self.connected = False
        self.connected_device = None 
        self.connection_enabled = True
        self.onConnection = True
		
		self.last_packet_time = datetime.now()
        self.rx_data = []
        self.rx_timestamp = []
        self.rx_delays = []

	def on_disconnect(self, client: BleakClient):
        self.connected = False
        self.onConnection = False
		
	def notification_handler(self, sender: str, data: Any):
        print(data)
        dataMap = data.decode("utf-8")
        if dataMap == "1":
            data_to_send = "Hello world!"
            self.rx_data.append(data_to_send)
            print(data_to_send)
        else: 
            data_to_send = "Hello people!"
            self.rx_data.append(data_to_send)
            print(data_to_send)
			
	async def manager(self):
        print("Starting connection manager.")
        while True:
            if self.client and self.onConnection==True:
                print("I am here and ready to be connected!")
                await self.connect()
            else:
                print("I am here and I am looking for new devices!")
                await self.select_device()
                await asyncio.sleep(1.0, loop=loop)
                self.onConnection = True
				
	async def connect(self):
	
        if self.onConnection==True:
            if self.connection_enabled:
                try:
                    await self.client.connect()
                    self.connected = await self.client.is_connected()
                    if self.connected:
                        print("Connected to Device")
                        self.client.set_disconnected_callback(self.on_disconnect)
                        await self.client.start_notify(
                            self.message_uuid, self.notification_handler,
                        )
                        while True:
                            if not self.connected:
                                break
                            print("From on_disconnect function I arrived here!")
                            await asyncio.sleep(1.0)
                    else:
                        print(f"Failed to connect to Device")
                except Exception as e:
                    print(e)
            else:
                await asyncio.sleep(1.0)
        else:
            await self.manager()
			
	async def cleanup(self):
        if self.client:

            await self.client.disconnected()
			
	async def select_device(self):
        print("Bluetooh LE hardware warming up...")
        await asyncio.sleep(2.0, loop=loop)
		self.client = BleakClient(address)  
	
	async def main():
    while True:

        # YOUR APP CODE WOULD GO HERE.
        await asyncio.sleep(1000)

        
         if __name__ == "__main__":

    os.environ["PYTHONASYNCIODEBUG"] = str(1)
    loop = asyncio.get_event_loop()

    data_to_file = DataToFile(output_file)

    connection = ArduinoConnection(loop, message_uuid, data_buffer_size, data_to_file.write_to_txt)

    try:
        asyncio.ensure_future(main())
        asyncio.ensure_future(connection.manager())
        loop.run_forever()
    except KeyboardInterrupt:
        print("User stopped the connection!")
    finally:
        print("Disconnecting....")
        loop.run_until_complete(connection.cleanup())
		
```

**Comment by dlech** [comment]
You need to use `BleakClient.get_services()` to enumerate the services to find the exact characteristic you want to use.

**Reply by dlech** [reply]
Related suggestion for a possible future alternate solution: #888

**Reply by ASabovic** [reply]
In my case, I already defined the message_uuid based on which it is able to receive data from my Arduino board. This function **BleakClient.get_services()** should be used after **self.client** (inside def select_device function in my case) is selected or?

**Reply by dlech** [reply]
It needs to be called after `await self.client.connect()`.

**Reply by ASabovic** [reply]
I added like this: 

**await self.client.connect()
self.client.get_services()**

But nothing really changed same error again. And also tried with BleakClient instead of self.client.get_service() but then it is not able to connect/receive anything at all.

**Reply by dlech** [reply]
You need to do more than that. You need to enumerate the services returned by this method to find the service you want, then enumerate that service to find the characteristic you want, then pass the characteristic object to `.start_notify()` instead of the UUID.

**Reply by ASabovic** [reply]
I found this possible solution that you posted before but there I am a bit confused: 

```
async with BleakClient(device) as client:
    my_service, = (s for s in client.get_services() if s.uuid == MY_SERVICE_UUID)
    my_char = my service.get_characteristic(MY_CHAR_UUID)
    client.write_gatt_char(my_char, MY_VALUE)

```
In this case, my service should be the BLEService defined on the Arduino side as well (for example Battery Service or something else). The next thing, my_char in this case is BLE Characteristics same as I defined in my code as message_uuid? And the last question, what is MY_VALUE here in your proposal? Is that the data that I am planning to receive or something else? Thanks a lot in advance!

**Reply by dlech** [reply]
You can use the `service_explorer.py` example program to print out the info on your device to find the service UUID and the characteristic UUID. In your previous code example, you were enabling notifications, not writing, so `MY_VALUE` is not relevant.

**Reply by ASabovic** [reply]
Sorry for this late reply, I solved the issue. Now everything works as I expected. Thanks a lot! :)
